[Carry 19133] Decouple server routers from the daemon package.

calavera opened the pull request:
Carry #19133.

cr7pt0gr4ph7 committed:
Decouple the "container" router from the actual daemon implementation.

This is done by moving the following types to api/types/config.go:
  - ContainersConfig
  - ContainerAttachWithLogsConfig
  - ContainerWsAttachWithLogsConfig
  - ContainerLogsConfig
  - ContainerStatsConfig

Remove dependency on "version" package from types.ContainerStatsConfig.
Decouple the "container" router from the "daemon/exec" implementation.

* This is done by making daemon.ContainerExecInspect() return an interface{}
value. The same trick is already used by daemon.ContainerInspect().

Improve documentation for router packages.
Extract localRoute and router into separate files.
Move local.router to image.imageRouter.

Changes:
  - Move local/image.go to image/image_routes.go.
  - Move local/local.go to image/image.go
  - Rename router to imageRouter.
  - Simplify imports for image/image.go (remove alias for router package).

Merge router/local package into router package.
Decouple the "image" router from the actual daemon implementation.
Add Daemon.GetNetworkByID and Daemon.GetNetworkByName.
Decouple the "network" router from the actual daemon implementation.

This is done by replacing the daemon.NetworkByName constant with
an explicit GetNetworkByName method.

Remove the unused Daemon.GetNetwork method and the associated constants NetworkByID and NetworkByName.

Signed-off-by: Lukas Waslowski <cr7pt0gr4ph7@gmail.com>
Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
Oh, I thought the changes in engine api were already merged :pensive:

calavera commented:
**DEPENDS ON** docker/engine-api#64

MHBauer commented:
Aw dang. I liked the individual commits.

I feel like doing the movement of the config structs right here is kind of weird, but if you can pull it off, go for it.

calavera committed:
Move backend types to their own package.

- Remove duplicated structs that we already have in engine-api.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
> I feel like doing the movement of the config structs right here is kind of weird, but if you can pull it off, go for it.

Agreed. I put them in a separated package here until we extract the whole server.

MHBauer commented:
Cool, I like it better like this. LGTM.

tiborvass commented:
if we could wait for 1.10 to be released before merging this that'd be really appreciated :) sorry!

anusha-ragunathan commented on api/types/backend/backend.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+// ContainerLogsConfig holds configs for logging operations. Exists
+// for users of the daemon to to pass it a logging configuration.
```
"daemon" -> "backend". Same in L56.

anusha-ragunathan commented on api/types/backend/backend.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+// ContainersConfig is the filtering specified by the user to iterate over containers.
+type ContainersConfig struct {
```
Looks similar to https://github.com/docker/engine-api/blob/master/types/client.go#L44

anusha-ragunathan commented on api/types/backend/backend.go:
```diff
@@ -0,0 +1,62 @@
+
+// ContainerLogsConfig holds configs for logging operations. Exists
+// for users of the daemon to to pass it a logging configuration.
+type ContainerLogsConfig struct {
```
Similar to https://github.com/docker/engine-api/blob/master/types/client.go#L56

anusha-ragunathan commented on api/types/backend/backend.go:
```diff
@@ -0,0 +1,62 @@
+
+// ContainerStatsConfig holds information for configuring the runtime
+// behavior of a daemon.ContainerStats() call.
+type ContainerStatsConfig struct {
```
It would be better to move this to types/client.go and rename to ContainerStatsOptions, like other client options.

anusha-ragunathan commented on api/server/server.go:
```diff
@@ -177,7 +177,7 @@ func (s *Server) makeHTTPHandler(handler httputils.APIFunc) http.HandlerFunc {
 // InitRouters initializes a list of routers for the server.
 func (s *Server) InitRouters(d *daemon.Daemon) {
```
This init function should take a backend. I'm fine if it is done in a follow on PR.

calavera replied:
I agree

calavera replied:
maybe we should remove this function in the future, or group all backends into a single `APIBackend`.

calavera commented:
@anusha-ragunathan as we mention in previous comments, we're not going to move these types to engine-api until we extract the whole server layer.

anusha-ragunathan commented:
In which case, a TODO comment there would be useful.

MHBauer commented:
Don't worry, I'll remember.

calavera commented:
I added a note at the beginning of the `backend` package about the future migration. I also removed duplicated structs that existed already in engine-api.

anusha-ragunathan commented:
Looks okay to me.

GordonTheTurtle commented:
Job: Docker-PRs FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs/23551/consoleText as the failures:

~~~console

---
)
Congratulations!  All Go source files have been linted.

---> Making bundle: validate-pkg (in bundles/1.11.0-dev/validate-pkg)
Congratulations! "./pkg/..." is safely isolated from internal code.

---> Making bundle: validate-test (in bundles/1.11.0-dev/validate-test)
Congratulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image.go:38: r.postImagesLoad undefined (type *imageRouter has no field or method postImagesLoad)
api/server/router/image/image_routes.go:272: undefined: router
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson7314018931410676534.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr23551
Error: failed to remove containers: [docker-pr23551]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14762/consoleText as the failures:

~~~console

---
)
Congratulations!  All Go source files have been linted.

---> Making bundle: validate-pkg (in bundles/1.11.0-dev/validate-pkg)
Congratulations! "./pkg/..." is safely isolated from internal code.

---> Making bundle: validate-test (in bundles/1.11.0-dev/validate-test)
Congratulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image.go:38: r.postImagesLoad undefined (type *imageRouter has no field or method postImagesLoad)
api/server/router/image/image_routes.go:272: undefined: router
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson5494012179675726768.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr-exp14762
Error: failed to remove containers: [docker-pr-exp14762]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1670/consoleText as the failures:

~~~console

---
 global name
 func (s *router) postImagesLoad(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
          ^
api/server/router/image/image.go:3:50: note: 'router' imported here
 import "github.com/docker/docker/api/server/router"
                                                  ^
api/server/router/image/image.go:38:40: error: reference to undefined field or method 'postImagesLoad'
   router.NewPostRoute("/images/load", r.postImagesLoad),
                                        ^
api/server/router/image/image_routes.go:272:10: error: use of undefined type 'router'
 func (s *router) postImagesLoad(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
          ^
api/server/router/image/image_routes.go:278:10: error: reference to field 'daemon' in object which has no fields or methods
  return s.daemon.LoadImage(r.Body, w, quiet)
          ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson4464251970275950106.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr-gccgo1670
Error: failed to remove containers: [docker-pr-gccgo1670]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-userns/5989/consoleText as the failures:

~~~console

---
-> 749d5567c576
Removing intermediate container 48fa6b312677
Successfully built 749d5567c576

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image.go:38: r.postImagesLoad undefined (type *imageRouter has no field or method postImagesLoad)
api/server/router/image/image_routes.go:272: undefined: router
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson6581293551893274611.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr-userns5989
Error: failed to remove containers: [docker-pr-userns5989]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/747/console):

~~~console

---
n.LoadImage
+ ec=1
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 1 at Mon Feb  8 17:37:02 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Mon Feb  8 17:37:02 CUT 2016 (1m 21s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson4833052116809128655.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14768/console):

~~~console

---
All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image_routes.go:278: too many arguments in call to s.daemon.LoadImage
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson7815906163577558247.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/5995/console):

~~~console

---
ng intermediate container 2935989eacba
Step 49 : COPY . /go/src/github.com/docker/docker
 ---> 61a731cc085f
Removing intermediate container 010397a45d26
Successfully built 61a731cc085f

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image_routes.go:278: too many arguments in call to s.daemon.LoadImage
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson1385079867603601489.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/749/console):

~~~console

---
800905e.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Mon Feb  8 17:52:57 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Mon Feb  8 17:52:58 CUT 2016 (1m 4s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson2583232419964128351.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1676/console):

~~~console

---
ARNING! I don't seem to be running in a Docker container.
# The result of this command might be an incorrect build, and will not be
# officially supported.
#
# Try this instead: make all
#

---> Making bundle: gccgo (in bundles/1.11.0-dev/gccgo)
# github.com/docker/docker/api/server/router/image
api/server/router/image/image_routes.go:278:17: error: too many arguments
  return s.daemon.LoadImage(r.Body, w, quiet)
                 ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson38174994960925029.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21521/console):

~~~console

---
ing bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev.exe
# github.com/docker/docker/api/server/router/image
api\server\router\image\image_routes.go:278: too many arguments in call to s.daemon.LoadImage
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson5348402091671686876.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23557/console):

~~~console

---
All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/image
api/server/router/image/image_routes.go:278: too many arguments in call to s.daemon.LoadImage
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson789477673746829698.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/750/console):

~~~console

---
8de8406.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Mon Feb  8 18:08:52 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Mon Feb  8 18:08:53 CUT 2016 (1m 5s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson8451611260190775696.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21522/console):

~~~console

---
ttachSilent	1.242s
PASS: docker_cli_start_test.go:112: DockerSuite.TestStartMultipleContainers	2.382s
PASS: docker_cli_start_test.go:96: DockerSuite.TestStartPausedContainer	0.933s
PASS: docker_cli_start_test.go:71: DockerSuite.TestStartRecordError	1.933s
PASS: docker_api_containers_test.go:1425: DockerSuite.TestStartWithNilDNS	0.695s
PASS: docker_api_containers_test.go:882: DockerSuite.TestStartWithTooLowMemoryLimit	0.161s

----------------------------------------------------------------------
FAIL: docker_cli_stats_test.go:98: DockerSuite.TestStatsAllNewContainersAdded

docker_cli_stats_test.go:131:
    c.Fatal("failed to observe new container created added to stats")
... Error: failed to observe new container created added to stats


----------------------------------------------------------------------
PASS: docker_cli_stats_test.go:80: DockerSuite.TestStatsAllNoStream	3.232s
PASS: docker_cli_stats_test.go:56: DockerSuite.TestStatsAllRunningNoStream	3.861s
PASS: docker_cli_stats_test.go:43: DockerSuite.TestStatsContainerNotFound	3.179s
PASS: docker_cli_stats
---

---
n
SKIP: docker_cli_push_test.go:281: DockerTrustSuite.TestTrustedPushWithoutServerAndUntrusted
SKIP: docker_cli_run_test.go:3139: DockerTrustSuite.TestTrustedRun
SKIP: docker_cli_run_test.go:3231: DockerTrustSuite.TestTrustedRunFromBadTrustServer
SKIP: docker_cli_create_test.go:309: DockerTrustSuite.TestUntrustedCreate
SKIP: docker_cli_pull_trusted_test.go:49: DockerTrustSuite.TestUntrustedPull
SKIP: docker_cli_run_test.go:3171: DockerTrustSuite.TestUntrustedRun
OOPS: 826 passed, 155 skipped, 1 FAILED
--- FAIL: Test (1823.76s)
FAIL
coverage: 49.1% of statements
exit status 1
FAIL	github.com/docker/docker/integration-cli	1824.171s
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson4122945027452
---

---
ker_cli_push_test.go:281: DockerTrustSuite.TestTrustedPushWithoutServerAndUntrusted
SKIP: docker_cli_run_test.go:3139: DockerTrustSuite.TestTrustedRun
SKIP: docker_cli_run_test.go:3231: DockerTrustSuite.TestTrustedRunFromBadTrustServer
SKIP: docker_cli_create_test.go:309: DockerTrustSuite.TestUntrustedCreate
SKIP: docker_cli_pull_trusted_test.go:49: DockerTrustSuite.TestUntrustedPull
SKIP: docker_cli_run_test.go:3171: DockerTrustSuite.TestUntrustedRun
OOPS: 826 passed, 155 skipped, 1 FAILED
--- FAIL: Test (1823.76s)
FAIL
coverage: 49.1% of statements
exit status 1
FAIL	github.com/docker/docker/integration-cli	1824.171s
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson412294502745274098.sh
+ set +x


---

---
81: DockerTrustSuite.TestTrustedPushWithoutServerAndUntrusted
SKIP: docker_cli_run_test.go:3139: DockerTrustSuite.TestTrustedRun
SKIP: docker_cli_run_test.go:3231: DockerTrustSuite.TestTrustedRunFromBadTrustServer
SKIP: docker_cli_create_test.go:309: DockerTrustSuite.TestUntrustedCreate
SKIP: docker_cli_pull_trusted_test.go:49: DockerTrustSuite.TestUntrustedPull
SKIP: docker_cli_run_test.go:3171: DockerTrustSuite.TestUntrustedRun
OOPS: 826 passed, 155 skipped, 1 FAILED
--- FAIL: Test (1823.76s)
FAIL
coverage: 49.1% of statements
exit status 1
FAIL	github.com/docker/docker/integration-cli	1824.171s
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson412294502745274098.sh
+ set +x


---

---
AndUntrusted
SKIP: docker_cli_run_test.go:3139: DockerTrustSuite.TestTrustedRun
SKIP: docker_cli_run_test.go:3231: DockerTrustSuite.TestTrustedRunFromBadTrustServer
SKIP: docker_cli_create_test.go:309: DockerTrustSuite.TestUntrustedCreate
SKIP: docker_cli_pull_trusted_test.go:49: DockerTrustSuite.TestUntrustedPull
SKIP: docker_cli_run_test.go:3171: DockerTrustSuite.TestUntrustedRun
OOPS: 826 passed, 155 skipped, 1 FAILED
--- FAIL: Test (1823.76s)
FAIL
coverage: 49.1% of statements
exit status 1
FAIL	github.com/docker/docker/integration-cli	1824.171s
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson412294502745274098.sh
+ set +x


---

~~~

tiborvass commented:
LGTM

MHBauer commented:
w00t w00t
